lib/Dialect/FIRRTL/Transforms/InferWidths.cpp

```diff
@@ -1149,8 +1154,62 @@ LogicalResult InferenceMapping::mapOperation(Operation *op) {
         // module's ports, and use them for instance port wires. This way,
         // constraints imposed onto the ports of the instance will transparently
         // apply to the ports of the instantiated module.
-        for (auto it : llvm::zip(module.getArguments(), op->getResults())) {
-          unifyTypes(FieldRef(std::get<0>(it), 0), std::get<1>(it));
+        for (auto it : llvm::zip(op->getResults(), module.getArguments())) {
+          unifyTypes(FieldRef(std::get<0>(it), 0), FieldRef(std::get<1>(it), 0),
+                     std::get<0>(it).getType().template cast<FIRRTLType>());
+        }
+      })
+
+      // Handle memories.
+      .Case<MemOp>([&](auto op) {
+        // Create constraint variables for all ports.
+        for (auto result : op.results())
+          declareVars(result, op.getLoc());
+
+        // This creates independent variables for every data port. Yet, what we
+        // actually want is for all data ports to share the same variable. To do
+        // this, we find the first data port declared, and use that port's vars
+        // for all the other ports.
+        unsigned firstFieldID;
+        switch (op.getPortKind(0)) {
+        case MemOp::PortKind::Read:
+          firstFieldID = 3;
+          break;
+        case MemOp::PortKind::Write:
+          firstFieldID = 3;
+          break;
+        case MemOp::PortKind::ReadWrite:
+          firstFieldID = 4;
+          break;
+        }
```

Comment:
The first two cases could be collapsed via fallthrough.

Going further, this whole thing can reduce to:

```c++
unsigned firstFieldID = 3;
if (op.getPortKind(0) == MemOp::PortKind::ReadWrite)
  firstFieldID = 4;
```

This is less clear than the `case` construction, so do whatever makes sense to you.

Reply:
Moved into a helper :+1:

Reply:
I like the static const + array ref solution. :100: